node/src/p2p.rs

```diff
@@ -105,65 +146,34 @@ impl P2p {
                 .map_err(P2pError::GossipsubInit)?;
 
         // subscribe to the topic
-        let header_sub_topic = gossipsub_ident_topic(&config.network_id, "/header-sub/v0.0.1");
+        let header_sub_topic = gossipsub_ident_topic(&args.network_id, "/header-sub/v0.0.1");
         gossipsub.subscribe(&header_sub_topic)?;
 
         let behaviour = Behaviour {
             identify,
             gossipsub,
-            header_ex: exchange::new_behaviour(&config.network_id),
+            header_ex: exchange::new_behaviour(&args.network_id),
             keep_alive: keep_alive::Behaviour,
         };
 
         let mut swarm =
-            SwarmBuilder::with_executor(config.transport, behaviour, local_peer_id, Executor)
-                .build();
+            SwarmBuilder::with_executor(args.transport, behaviour, local_peer_id, Executor).build();
 
-        for addr in config.listen_on {
+        for addr in args.listen_on {
             swarm.listen_on(addr)?;
         }
 
-        for addr in config.bootstrap_peers {
+        for addr in args.bootstrap_peers {
             swarm.dial(addr)?;
         }
 
-        let (cmd_tx, cmd_rx) = flume::bounded(16);
-
-        spawn(async move {
-            Worker {
-                cmd_rx,
-                swarm,
-                header_sub_topic_hash: header_sub_topic.hash(),
-            }
-            .run()
-            .await;
-        });
-
-        Ok(P2p { cmd_tx })
-    }
-
-    pub async fn next_event(&self) -> Option<P2pEvent> {
-        todo!();
+        Ok(Worker {
+            cmd_rx,
+            swarm,
+            header_sub_topic_hash: header_sub_topic.hash(),
+        })
     }
 
-    async fn send_cmd(&self, cmd: P2pCmd) -> Result<()> {
-        self.cmd_tx
-            .send_async(cmd)
-            .await
-            .map_err(|_| P2pError::WorkerDied)
-    }
-
-    pub async fn network_info(&self) -> Result<NetworkInfo> {
-        let (tx, rx) = oneshot::channel();
-
-        self.send_cmd(P2pCmd::NetworkInfo { respond_to: tx })
-            .await?;
-        rx.await
-            .map_err(|_| P2pError::CmdCancelled("NetworkInfo".to_owned()))
-    }
-}
-
-impl Worker {
     async fn run(&mut self) {
         let mut command_stream = self.cmd_rx.clone().into_stream().fuse();
```

Comment:
I didn't notice this in the previous PR. I think I prefer `self.cmd_rx.recv_async()`

Reply:
I couldn't find any info whether `recv_async` is cancel safe tho

Reply:
It is cancel safe but not documented, we can keep the fused stream for now.